llvm/docs/GlobalISel/MIRPatterns.rst

```diff
@@ -514,3 +514,29 @@ of operands.
     (match (does_not_bind $tmp, $x)
            (G_MUL $dst, $x, $tmp)),
     (apply (COPY $dst, $x))>;
+
+
+ Gallery
+----------------
+
+We should use precise patterns that state our intentions. Please avoid
+using wip_match_opcode in patterns.
+
+.. code-block:: text
+  :caption: Example fold ext(trunc:nuw)
+
+(match (wip_match_opcode G_ZEXT):$root,
+   [{ return Helper.matchZextOfTrunc(${root}, ${matchinfo}); }]),
+   (apply [{ Helper.applyBuildFnMO(${root}, ${matchinfo}); }])>;
+
+
+(match (G_TRUNC $src, $x),
```

Comment:
I think it's a good idea to have some examples. I'd like it even more if each one had a comment explaining what it matches, either in English or in some pseudocode like `Match (zext (trunc x))`.

Reply:
At the top it states: fold ext(trunk:nuw)
I want to prevent misbehaviour like:
https://github.com/llvm/llvm-project/blob/fc538b070de69315fe05105e8c0bc622fb61168a/llvm/include/llvm/Target/GlobalISel/Combine.td#L722

Reply:
What is the misbehaviour?

Reply:
It registers on G_ZEXT:
```
(match (wip_match_opcode G_ZEXT):$root,
```
but it states as a comment that it wants to fold zext(trunk(x).

Reply:
What's wrong with that? wip_match_opcode just matches the "topmost" / "outermost" opcode, which is zext in this case. The C++ code looks for the inner node, trunc.

Reply:
It will match *any* G_ZEXT in the function. In most cases, there is no trunk on the zext. It is a waste of compile-time. If we would write more precise matches, it might reduce compile-time.

wip_match_opcode should be forbidden. Please write precise patterns.

Reply:
OK. I was confused because "misbehaviour" generally implies that something is incorrect, not just suboptimal.

I would prefer comments on the individual patterns, e.g.:
```
// Imprecise: matches any G_ZEXT
...
// Better: matches G_ZEXT of G_TRUNC
...
// Precise: matches G_ZEXT of G_TRUNC with nuw flag
...
```
Without that, it is not clear to me why you have presented a list of similar-but-different patterns.